<a href="https://colab.research.google.com/github/heysatya/DeepLearningKeras/blob/main/Graded_Assignment_Project_Regression_Model_in_Keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align=center><font size = 5>Graded Assignment Project - Regression Model in Keras</font></h1>


## Introduction


In this course project, I will build a regression model using the deep learning Keras library, and then experiment with increasing the number of training epochs and changing number of hidden layers to find out how changing these parameters impacts the performance of the model.


<h4>Objective for this Assignment<h4>    
<h5> A. Build a baseline model </h5>
<h5> B. Normalize the data </h5>
<h5> C. Increase the number of epochs </h5>
<h5> D. Increase the number of hidden layers </h5>   


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
<a href="#item31">A. Build a baseline model</a>  
<a href="#item32">B. Normalize the data</a>  
<a href="#item33">C. Increase the number of epochs </a>  
<a href="#item34">D. Increase the number of hidden layers </a>  

</font>
</div>


<a id="item31"></a>


## A. Build a baseline model

In this first step, we will use the Keras library to build a neural network with the following:
- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.


#### Import the Pandas and the Numpy libraries.


In [1]:
import pandas as pd
import numpy as np

The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. 
    
The data has been downloaded from: https://cocl.us/concrete_data.
    
The predictors in the data of concrete strength include include:
1. Cement
2. Blast Furnace Slag
3. Fly Ash
4. Water
5. Superplasticizer
6. Coarse Aggregate
7. Fine Aggregate


#### Read the data into a Pandas dataframe.


In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

As can be seen from the data, the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. This concrete mix is 28 days old and has a compressive strength of 79.99 MPa. 


#### Split data into predictors and target


The target variable in this case is the concrete sample strength. Therefore, all the columns will be considered as predictors.


In [3]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>


#### Sanity check of the predictors and the target dataframes.


In [4]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [5]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

#### Save the number of predictors to _n_cols_ since we will need this number later on.


In [6]:
n_cols = predictors.shape[1] # number of predictors

#### Import the Keras library


In [7]:
import keras

#### Import the rest of the packages from the Keras library that we will need to build the regressoin model.


In [8]:
from keras.models import Sequential
from keras.layers import Dense

### Build a Neural Network


Define a function that defines the regression model that can be conveniently called it to create our model. Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.


In [9]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

#### 1. Randomly split the data into a training and test sets by holding 30% of the data for testing, using the train_test_split helper function from scikit-learn

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)

#### Create the model.


In [12]:
# build the model
model = regression_model()

#### 2. Train the model on the training data using 50 epochs.

In [13]:
model.fit(Xtrain, ytrain, epochs=50, verbose=1)

Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 14600.0227
Epoch 2/50
23/23 [==============================] - 0s 966us/step - loss: 10377.2459
Epoch 3/50
23/23 [==============================] - 0s 985us/step - loss: 7820.3005
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 5724.2302
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 4142.0964
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 3311.8290
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 2530.4533
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 1744.2256
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 1077.7043
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 792.4590
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 556.8784
Epoch 12/50
23/23 [==============================] - 0s 934us/step - loss: 470.3434
Epoch 13

#### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength, using the mean_squared_error function from scikit-learn.

In [14]:
model.evaluate(Xtest, ytest)
ypred = model.predict(Xtest)

10/10 [==============================] - 0s 2ms/step - loss: 186.7337


In [15]:
from sklearn.metrics import mean_squared_error

In [16]:
MSE = mean_squared_error (ytest, ypred)
print ("Part A: MSE between predicted and actual concrete strength: ", MSE)
print ("Part A: Mean of MSE: ", MSE.mean())
print ("Part A: Standard Deviation of MSE: ", MSE.std())

Part A: MSE between predicted and actual concrete strength:  186.7337131315141
Part A: Mean of MSE:  186.7337131315141
Part A: Standard Deviation of MSE:  0.0


#### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors

In [17]:
MSE = np.empty(50)
for i in range(0, 50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    model.evaluate(Xtest, ytest, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))

print ("Part A: List of Mean Squared Errors: ", MSE)

Part A: List of Mean Squared Errors:  [114.33344476 114.14777291 114.59100549 109.46981283 116.87766389
 121.58588691  96.5028271   99.04617639 118.54645696 102.68698379
  98.34339696 132.01336206 101.21486464 115.44651482 100.36476637
 114.38162294 105.41999222 102.79106567 116.62029787 113.88961578
 112.79021644 121.20319916 106.58513988 108.45108361 106.92059388
 103.15528988 114.09598156 104.74677739 119.03940636 104.21315859
 110.997756   113.65564806 105.31130776 108.40232174 107.23166329
 105.37824324 114.63211689 125.13753353 107.72651328 105.06445419
 131.01658442 116.30531882 101.78141039 100.32873688 116.09938977
 119.02049294 108.90269923 108.65728989 117.97887273 103.59229446]


#### 5. Report the mean and the standard deviation of the mean squared errors.

In [18]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part A: Mean of the 50 Mean Squared Errors: ", MSE_mean)
print ("Part A: Standard Deviation of the 50 Mean Squared Errors: ", MSE_std)

Part A: Mean of the 50 Mean Squared Errors:  110.73390049278207
Part A: Standard Deviation of the 50 Mean Squared Errors:  7.990016934860439


## B. Normalize the data

Repeating Part A using a normalized version of the data

Q) How does the mean of the mean squared errors compare to that from Step A?


#### Normalize the data by substracting the mean from the individual predictors and dividing by the standard deviation.

In [19]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

In [20]:
n_cols = predictors_norm.shape[1] # number of predictors

#### Build the model for Part B using normalized data

In [21]:
model = regression_model()

#### Repeat steps in Part A to create a list of 50 mean squared errors (using normalized data model)

In [22]:
MSE = np.empty(50)
for i in range(0, 50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors_norm, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    model.evaluate(Xtest, ytest, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part B: List of Mean Squared Errors: ", MSE)

Part B: List of Mean Squared Errors:  [335.16601754 164.4065708  109.10961777  92.60421439  75.73026168
  62.87715279  51.66579766  57.69869275  46.57561286  44.76744054
  43.52966197  40.70602564  47.72046752  39.63163973  43.6019422
  40.27716873  39.60446266  37.17717656  35.98524049  39.84953993
  36.68459025  36.72232214  37.2800111   36.91935267  34.81939211
  38.15697477  33.79150147  37.63538207  39.24147501  35.4859834
  37.62477986  34.14268836  34.82477785  39.56784215  39.19330583
  39.96070138  33.61883541  31.51946425  32.21166128  35.7402847
  37.2551656   31.96667489  39.57895787  29.53174818  37.78674783
  37.09662492  31.76475727  28.81488299  32.56886789  34.22736348]


#### Calculating mean and standard deviation for Part B

In [23]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part B: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part B: Standard Deviation of the Mean Squared Errors: ", MSE_std)

Part B: Mean of the Mean Squared Errors:  50.28835642381871
Part B: Standard Deviation of the Mean Squared Errors:  46.50714455599244


Q) How does the mean of the mean squared errors compare to that from Step A?

A) Based on the results of Part A and Part B, the following can be observed:

> Part A: Mean of MSE = 110.733 & Std Dev of MSE  = 7.990

> Part B: Mean of MSE = 50.288 & Std Dev of MSE = 46.507

**As seen from the above results, the Mean of MSE for Part B is better than that of Part A.**


## C. Increase the number of epochs

Repeat Part B using 100 epochs this time for training.

Q) How does the mean of the mean squared errors compare to that from Step B?

#### Repeating steps in Part B using 100 epochs for training

#### Build the model for Part C

In [24]:
model = regression_model()

In [25]:
MSE = np.empty(50)
for i in range(0, 50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors_norm, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=100, verbose=0)
    model.evaluate(Xtest, ytest, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part C: List of Mean Squared Errors: ", MSE)

Part C: List of Mean Squared Errors:  [170.00017807  80.58591008  49.4981608   43.79580085  47.17233965
  37.94740124  35.11461423  36.50047266  35.49220912  38.53308485
  36.63699636  33.93757548  33.72578776  34.22623833  32.18227712
  33.31111364  31.06219919  31.06835568  33.02338326  26.76760629
  32.94710013  29.79052488  31.25973635  31.75470148  27.90326989
  32.41993981  28.40902302  32.18048273  25.76467962  27.68071382
  31.93056023  29.58299464  30.18173972  30.70254349  31.49813424
  29.60964349  24.68424734  28.15757718  29.80890607  27.61901618
  31.47085596  30.83555644  23.97563812  32.34251472  30.90511613
  34.50074698  30.21981503  25.29666691  29.4055441   29.49633449]


#### Calculating mean and standard deviation for Part C

In [26]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part C: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part C: Standard Deviation of the Mean Squared Errors: ", MSE_std)

Part C: Mean of the Mean Squared Errors:  35.858320557204586
Part C: Standard Deviation of the Mean Squared Errors:  20.908726583388717


Q) How does the mean of the mean squared errors compare to that from Step B?

A) Based on the results of Part B and Part C, the following can be observed:

> Part B: Mean of MSE = 50.288 & Std Dev of MSE = 46.507

> Part C: Mean of MSE = 35.858 & Std Dev of MSE = 20.908

As seen from the above results, the Mean of MSE for Part C is better than that of Part B.

## D. Increase the number of hidden layers

Repeat part B using a neural network with the following:

- Three hidden layers, each of 10 nodes and ReLU activation function.

Q) How does the mean of the mean squared errors compare to that from Step B?

#### Define regression model with 3 hidden layers, each of 10 nodes and ReLU activation functions


In [27]:
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

#### Build the model for Part D


In [28]:
model = regression_model()

#### Repeating steps in Part B using 50 epochs for training the neural network of Part D

In [29]:
MSE = np.empty(50)
for i in range(0, 50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors_norm, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    model.evaluate(Xtest, ytest, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part D: List of Mean Squared Errors: ", MSE)

Part D: List of Mean Squared Errors:  [125.99988435 116.78274604  93.98618267  74.36750922  55.97103409
  54.31866158  50.67216488  33.68769538  41.87529467  39.07856144
  42.63722795  31.79828134  35.447057    35.88231261  31.51626836
  28.70474366  39.30676608  28.04685344  28.66927502  30.27305184
  28.47783996  30.16737289  27.00416673  31.68935687  32.77580913
  22.70758444  26.53351517  21.0076048   27.54079018  26.45157084
  22.98750089  26.40317653  25.71459852  24.38551544  26.20387161
  28.14270092  24.96985376  27.33344295  25.82989653  20.79056899
  23.38863548  20.78390811  22.76515278  25.94880683  23.68372962
  23.66238265  23.15872589  19.8328555   20.8388503   22.81868064]


#### Calculating mean and standard deviation for Part D

In [30]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part D: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part D: Standard Deviation of the Mean Squared Errors: ", MSE_std)

Part D: Mean of the Mean Squared Errors:  35.46040073114769
Part D: Standard Deviation of the Mean Squared Errors:  22.175233527120138


Q) How does the mean of the mean squared errors compare to that from Step B?

A) Based on the results of Part B and Part D, the following can be observed:

> Part B: Mean of MSE = 50.288 & Std Dev od MSE = 46.507

> Part D: Mean of MSE = 35.460 & Std Dev of MSE = 22.175

As seen from the above results, the Mean of MSE for Part D is better than that of Part B, which can be attributed to the addition of another hidden layer to the neural network.